src/slist.c

```diff
@@ -1 +1,324 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdbool.h>
+
+#include "macro_utils/macro_utils.h"
+
+#include "c_logging/xlogging.h"
+
 #include "c_util/slist.h"
+
+bool slist_initialize(PSINGLYLINKEDLIST_ENTRY list_head)
+{
+	bool result;
+	if (list_head == NULL)
+	{
+		/* Codes_SRS_SLIST_07_001: [ If `list_head` is `NULL`, `slist_initialize` shall fail and return `false`. ]*/
+		LogError("Invalid argument (list_head=%p)", list_head);
+		result = false;
+	}
+	else
+	{
+		/* Codes_SRS_SLIST_07_002: [ `slist_initialize` shall initialize the `next` pointer in `list_head` points to `NULL`and return `true` on success. ]*/
+		list_head->next = NULL;
+		result = true;
+	}
+	return result;
+}
+
+bool slist_is_empty(const PSINGLYLINKEDLIST_ENTRY list_head)
+{
+	bool result;
+	if (list_head == NULL)
+	{
+		/* Codes_SRS_SLIST_07_038: [ If `list_head` is `NULL`, `slist_is_empty` shall fail and return `true`. ]*/
+		LogError("Invalid argument (list_head=%p)", list_head);
+		result = true;
+	}
+	else
+	{
+		if (list_head->next == NULL)
+		{
+			/* Codes_SRS_SLIST_07_003: [ `slist_is_empty` shall return `true` if there is no `SINGLYLINKEDLIST_ENTRY` in this list. ]*/
+			result = true;
+		}
+		else
+		{
+			/* Codes_SRS_SLIST_07_004: [ `slist_is_empty` shall return `false` if there is one or more entris in the list. ]*/
+			result = false;
+		}
+	}
+	return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_add(PSINGLYLINKEDLIST_ENTRY list_head, PSINGLYLINKEDLIST_ENTRY list_entry)
+{
+	PSINGLYLINKEDLIST_ENTRY result;
+	if ((list_head == NULL) ||
+		(list_entry == NULL))
+	{
+		/* Codes_SRS_SLIST_07_005: [ If `list_head` is `NULL`, `slist_add` shall fail and return `NULL`. ]*/
+		/* Codes_SRS_SLIST_07_006: [ If `list_entry` is `NULL`, `slist_add` shall fail and return `NULL`. ]*/
+		LogError("Invalid argument (list_head=%p, list_entry=%p)", list_head, list_entry);
+		result = NULL;
+	}
+	else
+	{
+		/* Codes_SRS_SLIST_07_007: [ `slist_add` shall add one entry to the tail of the list on success and return a pointer to the added entry. ]*/
+		PSINGLYLINKEDLIST_ENTRY list_instance = list_head;
+
+		while (list_instance->next != NULL)
+		{
+			list_instance = list_instance->next;
+		}
+
+		list_instance->next = list_entry;
+		result = list_entry;
+	}
+	return result;
+}
+
+PSINGLYLINKEDLIST_ENTRY slist_add_head(PSINGLYLINKEDLIST_ENTRY list_head, PSINGLYLINKEDLIST_ENTRY list_entry)
```

Comment:
>PSINGLYLINKEDLIST_ENTRY [](http://example.com/codeflow?start=0&length=23)

I don't think returning the list_entry here helps. It actually seems to make things more confusing as I was wondering really what are we returning when reading the tests.

Reply:
changed to return int here.